lib/ansible/module_utils/facts/hardware/linux.py

```diff
@@ -275,6 +275,21 @@ def get_cpu_facts(self, collected_facts=None):
                 cpu_facts['processor_vcpus'] = (cpu_facts['processor_threads_per_core'] *
                                                 cpu_facts['processor_count'] * cpu_facts['processor_cores'])
 
+                if sys.version_info[0] >= 3:
+                    cpu_facts['processor_nproc'] = len(
+                        os.sched_getaffinity(0)
```

Comment:
It's great that Python has this capability, but it's not always avaialble. In this conext (Python >=3 on Linux systems) it _should_ be available, but it's probably best to catch `AttributeError` just in case it's not available.

Reply:
Hm, I'd find it odd to not trust in what the docs of the standard library say. The docs do not state platform-dependent availability of the method.
But I totally see your point and checked when this method has been introduced. In 3.3. ([search in 3.2.6 docs with no results](https://docs.python.org/release/3.2.6/search.html?q=affinity), [search in 3.3.0 docs with results](https://docs.python.org/release/3.3.0/search.html?q=affinity)) and made the version check more accurate. Hence, I couldn't use ``PY3`` or is there another way to check for minor version numbers using ``six``?

Reply:
> The docs do not state platform-dependent availability of the method.

In particular, this isn't available on macOS. I found this caveat in the Python docs [here](https://docs.python.org/3/library/os.html#interface-to-the-scheduler):
> They are only available on some Unix platforms.

Again, probably not a big deal since Ansible should only be using this file on Linux where it should be available. But I like to be overly cautious.

> ...introduced. In 3.3. ... and made the version check more accurate.

Ansible only supports Python >= 3.5, so testing for `PY3` is sufficient.

Reply:
I think this approach is good. Exceptions in Python are not as expensive as in other languages so I don't see a problem with this approach. I would suggest reworking this a bit to make it easier for the reader, namely returning early when possible and eliminating the need for a conditional.

```python
                try:
                     cpu_facts['processor_nproc'] = len(
                         os.sched_getaffinity(0)
                     )
                    return cpu_facts
                except AttributeError:
                    # in Python < 3.3 and MacOS (shouldn't happen here,
                    # but to be safe) ``sched_getaffinity`` is not available
                    try:
                        cmd = get_bin_path('nproc')
                    except ValueError:
                        cpu_facts['processor_nproc'] = processor_occurence
                        return cpu_facts

                    rc, out, _err = self.module.run_command(cmd)
                    if rc == 0:
                        cpu_facts['processor_nproc'] = int(out)
                    else:
                        cpu_facts['processor_nproc'] = processor_occurence
```

The `cpu_facts.setdefault()` line can be omitted with this approach. This may be more lines of code, but I think it's a bit easier to follow. Just my opinion.

Reply:
Moving the assignment of the default value up is fine with me.
"Replacing" the `else` of the `try … except` with a `return` wouldn't increase readability for me.
Also, if code needs to be added below the computation of `processor_nproc`, it is probably easier if we do not return in corner cases but keep going.
But if you insist, I wouldn't veto.

Reply:
Ok, let's move the default assignment up and keep the `else` without returning early. The main way I think this should read is "start with `processor_occurence` and use something better if we can".

Reply:
> So just as the PR is right now, no?

Is this resolved? Any obligations left for a merge?